iothub_client/samples/pnp/pnp_simple_thermostat/pnp_simple_thermostat.c

```diff
@@ -0,0 +1,484 @@
+        // something is fundamentally wrong with how we've received the twin and we should not proceed.
+        LogError("JSON field %s is not a number but must be", PnP_JsonPropertyVersion);
+    }
+    else
+    {
+        double targetTemperature = json_value_get_number(targetTemperatureValue);
+        int version = (int)json_value_get_number(versionValue);
+
+        LogInfo("Received targetTemperature = %f", targetTemperature);
+
+        bool maxTempUpdated = false;
+        UpdateTemperatureAndStatistics(targetTemperature, &maxTempUpdated);
+
+        // The device needs to let the service know that it has received the targetTemperature desired property.
+        SendTargetTemperatureReport(deviceClient, targetTemperature, g_statusSuccess, version, g_temperaturePropertyResponseDescription);
+
+        if (maxTempUpdated)
+        {
+            // If the Maximum temperature has been updated, we also report this as a property.
+            SendMaxTemperatureSinceReboot(deviceClient);
+        }
+    }
+
+    json_value_free(rootValue);
+    free(jsonStr);
+}
+
+//
+// Thermostat_SendCurrentTemperature sends a PnP telemetry indicating the current temperature
+//
+void Thermostat_SendCurrentTemperature(IOTHUB_DEVICE_CLIENT_HANDLE deviceClient) 
+{
+    IOTHUB_MESSAGE_HANDLE messageHandle = NULL;
+    IOTHUB_CLIENT_RESULT iothubResult;
+
+    char temperatureStringBuffer[32];
+
+    if (snprintf(temperatureStringBuffer, sizeof(temperatureStringBuffer), g_TemperatureTelemetryBodyFormat, g_currentTemperature) < 0)
+    {
+        LogError("snprintf of current temperature telemetry failed");
+    }
+    else if ((messageHandle = IoTHubMessage_CreateFromString(temperatureStringBuffer)) == NULL)
+    {
+        LogError("IoTHubMessage_CreateFromString failed");
+    }
+    else if ((iothubResult = IoTHubDeviceClient_SendEventAsync(deviceClient, messageHandle, NULL, NULL)) != IOTHUB_CLIENT_OK)
+    {
+        LogError("Unable to send telemetry message, error=%d", iothubResult);
+    }
+
+    IoTHubMessage_Destroy(messageHandle);
+}
+
+int main(void)
+{
+    IOTHUB_DEVICE_CLIENT_HANDLE deviceClient = NULL;
+    const char* connectionString;
+
+    if ((connectionString = getenv(g_ConnectionStringEnvironmentVariable)) == NULL)
+    {
+        LogError("Cannot read environment variable %s", g_ConnectionStringEnvironmentVariable);
+    }
+    else if (BuildUtcTimeFromCurrentTime(g_ProgramStartTime, sizeof(g_ProgramStartTime)) == false)
+    {
+        LogError("Unable to output the program start time");
+    }
+    else if ((deviceClient = PnPHelper_CreateDeviceClientHandle(connectionString, g_ModelId, g_hubClientTraceEnabled, Thermostat_DeviceMethodCallback, Thermostat_DeviceTwinCallback)) == NULL)
+    {
+        LogError("Failure creating Iothub device");
+    }
+    else
+    {
+        LogInfo("Successfully created device client handle.  Hit Control-C to exit program\n");
```

Comment:
`desired.targetTemperature` should be read and ACK at startup

Reply:
This does indeed happen.  C's `IoTHubDeviceClient_SetDeviceTwinCallback` does an implicit GetTwin in addition to subscribing to PATCH updates.